### ISSUE TITLE
Reference Error when using exists() method with node.js

### ISSUE THREAD

**Issue by jddahl** [anywhichway/rule-reactor#18]
I am implementing the library in a node.js application.
Trying to use the `exists()` method in the testing part of a rule, I systematically get an error  "ReferenceErrror: /name of object/ is not defined".

`    function(a) {
        return (not(exists({a:A},function(a){return a.i==true})));
        },` 

Using any other kind of tests works.
`    function(a) {
        return (a.i==false);
        },` 

_It is interesting that the same script implemented in a browser works._
So, it seems there is a behavior difference beween node and browser.

- Did you already experience such problems by node.js ?
- Is there a workaround ?

Many thanks.
JD DAHL

Node version is 6.9.5
Rule-Reactor version is 0.1.8

**Call stack:**
`ReferenceError: A is not defined
    at eval (eval at <anonymous> (C:\$JDD\nodeJS\git-yooz\audit-trail-analyzer\node_modules\rule-reactor\index.js:264:19), <anonymous>:2:11)
    at C:\$JDD\nodeJS\git-yooz\audit-trail-analyzer\node_modules\rule-reactor\index.js:264:54
    at RegExp.[Symbol.replace] (native)
    at RegExp.[Symbol.replace] (native)
    at String.replace (native)
    at C:\$JDD\nodeJS\git-yooz\audit-trail-analyzer\node_modules\rule-reactor\index.js:262:19
    at Array.forEach (native)
    at compile (C:\$JDD\nodeJS\git-yooz\audit-trail-analyzer\node_modules\rule-reactor\index.js:261:19)
    at new Rule (C:\$JDD\nodeJS\git-yooz\audit-trail-analyzer\node_modules\rule-reactor\index.js:510:3)
    at RuleReactor.createRule (C:\$JDD\nodeJS\git-yooz\audit-trail-analyzer\node_modules\rule-reactor\index.js:979:25)`

**Attached files in ZIP are:**
-testReactor3.html => working version in browser
-testReactor3.js => version having issue by node.js

[testReactor.zip](https://github.com/anywhichway/rule-reactor/files/848168/testReactor.zip)

**Comment by anywhichway** [comment]
@jddahl ... have not looked at your code yet. That being said, we frequently experience differences between a browser and node.js that are similar in nature to the type of thing you are describing in addition to major performance differences (browsers are almost always faster with a similar memory and CPU profile). The first suggestion is to try and upgrade to node 7.x.x and see if the problem goes away. Meanwhile, we will poke around in your code.

**Comment by anywhichway** [comment]
@jddahl Right above line 264 in index.js insert `console.log(domainstr,conditionstr)`  also insert `console.log(args)` right above line 350 and compare the results between the browser run and the node.js run.

```
function(match,domainstr,conditionstr) {
        console.log(domainstr,conditionstr);
	var domain = new Function("return " + domainstr)(), variables = Object.keys(domain);
	var quantification = {domain: domain, range: {}};
	rule.triggers.push(quantification);
```

```
})) {
  console.log(args);
  return condition.apply(reactor,args);
}
```

**Comment by jddahl** [comment]
Hi Simon,
Many thanks for your help.
I've put both traces ('*A*' above 264 and '*B* above 350->351) into
index.js and browser\rule-reactor.js.

By node and browser, both *A* traces write the content of the call to the
function exists(): "{a:A},function(a){return a.i==true}"
By node, *B* doesn't appear at all, and I get a "undefined:2" trace I don't
know where from it can come.
By browser, *B* writes several times the [A] object, ( 5 times before rule
"R1" and once before rule "end") (see log attached)

hoping this provides you some clues,
regards,
Jean-Dag Dahl
(from France)

**Comment by jddahl** [comment]
Hi,

In order to be sure, I made a test of the same simple script using node 7.7.3 version under a docker environment.
Exactly the same issue appears: 
`ReferenceError: A is not defined` line 264 of rule-reactor/index.js 

So that would seem related to any node.js implementation although not occurring on a browser.
Have you any hints ?
many thanks
JD Dahl

**Comment by anywhichway** [comment]
@jddahl  still digging into this. Hope to have a solution by this weekend.

**Comment by Zelgadis87** [comment]
@anywhichway Encountered the same issue on Node. Let me know if I can be of help.

**Comment by anywhichway** [comment]
@jddahl We have found the source of the issue. @Zelgadis87 is correct the "compiler" for exists is relying on A being defined within its scope. In the browser the rules are globally available. The fix will be to provide an import function for rule reactor, e.g. something like `reactor.declare(domain,constructor)`. Working on this.

**Comment by anywhichway** [comment]
@jddahl We have a fix. It is a bit dirty, but I don't know how urgent your need is, We are looking for something cleaner. However, I describe the fix below which works with the example code you sent:

1) Insert this change around line 15 of node_modules/rule-reactor/index.js

```(function() {
	"use strict";
	var domains = global;
```

2) Add this to the RuleReactor class:

```
RuleReactor.prototype.declare = function(domain,constructor) {
		if(typeof(module)==="object") {
			domains[domain] = constructor;
		}
	}
```

Let us know if this works. We will look to making a proper patch in the next week after noodling on the best fix approach some more.

**Comment by jddahl** [comment]
Hello,
Thanks for your reply.
Maybe I do something wrong, but this doesn't seem to work.

I put the change into index.js
```
/* MIT License

rule-reactor: A light weight, fast, expressive forward chaining business rule engine leveraging JavaScript internals and Functions as objects rather than Rete.

Copyright (c) 2016 Simon Y. Blackwell

*/

var uuid = require("uuid");
//var hamsters = require("webhamsters/src/hamsters");
//var Parallel = require("paralleljs");

(function() {
	"use strict";
	
// https:// github.com / anywhichway / rule-reactor / issues / 18
// proposed fix	
	var domains = global;			// FIX
// end of fix
```

and in my test script:
```

// https:// github.com / anywhichway / rule-reactor / issues / 18
// proposed fix
RuleReactor.prototype.declare = function(domain,constructor) {
		if(typeof(module)==="object") {
			domains[domain] = constructor;
		}
	}
// end of fix
	
	// works:
	//return (a.i==false);
	
	// doesn't work => raises "ReferenceError: A is not defined"
	//return (not(exists({a:A},function(a){return a.i==true})));
	
reactor.createRule("R1",1000,{a:A},
    function(a) {
        return (not(exists({a:A},function(a){return a.i==true})));
        },
    function(a) {
        a.i = true ;
        console.log('==> ',"R1");
        });
```

See the sources: 
[rule-reactor_issues_18.zip](https://github.com/anywhichway/rule-reactor/files/885087/rule-reactor_issues_18.zip)

However, I still get the same error
```
C:\$JDD\nodeJS\git-yooz\oldies\test-rule-engine\app>node testReactor3.js
*A* {a:A} function(a){return a.i==true}))
undefined:2
return {a:A}
          ^

ReferenceError: A is not defined
    at eval (eval at <anonymous> (C:\$JDD\nodeJS\git-yooz\oldies\test-rule-engine\node_modules\rule-reactor\index.js:270:19), <anonymous>:2:11)
    at C:\$JDD\nodeJS\git-yooz\oldies\test-rule-engine\node_modules\rule-reactor\index.js:270:54
    at RegExp.[Symbol.replace] (native)
    at RegExp.[Symbol.replace] (native)
    at String.replace (native)
    at C:\$JDD\nodeJS\git-yooz\oldies\test-rule-engine\node_modules\rule-reactor\index.js:267:19
    at Array.forEach (native)
    at compile (C:\$JDD\nodeJS\git-yooz\oldies\test-rule-engine\node_modules\rule-reactor\index.js:266:19)
    at new Rule (C:\$JDD\nodeJS\git-yooz\oldies\test-rule-engine\node_modules\rule-reactor\index.js:517:3)
    at RuleReactor.createRule (C:\$JDD\nodeJS\git-yooz\oldies\test-rule-engine\node_modules\rule-reactor\index.js:986:25)

```
Currently, I use a workaround in the rules I am developing by managing the knowledge of the existence of fact instances using some properties referencing them in other objects of the domain. This is not very nice as it makes the code heavier and less maintainable and I guess more prone to bugs.

Please notice that I have now personally to stop working on this project but it will be transferred to a colleague of mine.

Hence, this issue is now not so urgent but maybe you'll be contacted again by him in a few days or weeks.

Many thanks for your help.

**Comment by anywhichway** [comment]
@jddahl I am sorry, we left one one instruction. Before you use A in your rule, insert the following: 

```
reactor.declare("A",A);
```

i.e.

```
reactor.declare("A",A);
reactor.createRule("R1",1000,{a:A}, ...
```

**Comment by anywhichway** [comment]
@jddahl Fixed in v0.1.9 ... however, we have also know push v0.1.10
